### DISCUSSION TITLE
Catch DOM mutation event

### DISCUSSION THREAD

**Discussion by guiiamorim** [alpinejs/alpine#3222] [1. Help]
Hello,

I'm using a javascript select plugin that renders the select options inside another div. I'm loading the select options in alpine, and it works just fine on page load.

When a user performs an action (e.g,: selecting an option on another select), I need to filter this options and show only some of them. When I filter the options, the DOM is updated accordingly, but I must re-render the plugin to reflect the changes. The problem is that I have many selects that should be changed, so I'd like to avoid doing it manually.

I was hoping theres a way to hook into the mutation of the DOM by alpine so I could refresh the select's plugin once the options are updated.

This is what the DOM looks like when I update the options, the real select is empty, but the plugin's select is still populated.
![image](https://user-images.githubusercontent.com/9733153/197214368-69b20a46-dd60-4344-a621-9b9a3ecee6ab.png)

I'm using a directive to interact with the plugin, I was hoping I could do something inside this code, but if there's some global hook I can use in the alpine object or something else that I can use please let me know.
```JS
Alpine.directive('tail-select', (el, _, {effect}) => {
    inputMasks['tail-select'](el) // This instantiate the plugin on the select (I'm using alpine on an existing project)
    setTimeout(() => el._tail.reload(), 250) // I need this timeout to get the options loaded by alpine on init

    effect(() => {
        el._tail?.reload() // Tried to see if when the options changed the effect would execute, but nothing happens
    })
    el._tail.on('change', evt => {
        el._x_model.set(el._tail.value()) // Needed this because the plugin was screwing my model binding
    })
})
```

**Comment by SimoTod** [comment]
So, the way effect works is this: 
- it accepts a callback
- that callback runs once
- if during the code execution it touches an alpine variable (something defined in x-data, technically something wrapped in a special proxy by alpine), Alpine will trigger the callback again when that variable changes.

In your code, effect doesn't track anything, you are just accessing the third party library. What you want to do is to pass the variable to track to tour custom directive (the same you use to generate your list). 

Inside the directive, you can evaluate then evaluate it making sure you touch the variable inside the effect callback.

I suspect the code can also be simplified. The effect itself should go into the timeout so you just call reload once

**Reply by guiiamorim** [reply]
I think I didn't understand your explanation, I changed the code to use evaluateLater and it still doesn't work.

```JS
Alpine.directive('tail-select', (el, {expression}, {evaluateLater, effect}) => {
    const getOptions = evaluateLater(expression)
    let lastOptions
    inputMasks['tail-select'](el)
    setTimeout(() => el._tail.reload(), 250)

    effect(() => {
        getOptions(options => {
            if (lastOptions === undefined) {
                lastOptions = options
            }

            if (options != lastOptions) {
                el._tail?.reload()
                lastOptions = options
            }
        })
    })
    el._tail.on('change', evt => {
        el._x_model.set(el._tail.value())
    })
})
```

What am I doing wrong?

**Reply by SimoTod** [reply]
I suppose tail-select needs to wait for the select to be updated. Directives are processed in order, when your callbaxk triggers, x-for hasn't run yet.

**Reply by guiiamorim** [reply]
But the thing is, if I put a console.log inside the effect, even when I change the list (inside alpine), nothing gets logged. I would expect that at least the log appeared in the console, but appears that the function just runs once on initialization

**Reply by SimoTod** [reply]
Not sure, your code has some checks that I don't think are necessary. Assuming tail_select is https://github.com/wolffe/tail.select.js, your directive should be something like https://codepen.io/SimoTod/pen/NWMQYvb (I added some comments to explain the critical bits that I suppose were failing for you, I've also removed the timeout which is not a robust way to wait for something to happen)

**Reply by guiiamorim** [reply]
I discovered what was preventing the effect from running, I was updating the array directly, not setting a new value on the variable. I'm still new to alpine, so I get confused with some aspects of it. I will make some modifications to my code following what you provided in the pen, thank you very much!

**Comment by guiiamorim** [comment]
Hi @SimoTod,

Sorry for bothering you again with this, but I discovered another problem with my solution and would like to know if you have any idea how I could solve this.

My problem this time is that once I update the model which the select is bound, the real select is updated accordingly, but the tail select instance is not updated. I suppose that if I put the model in the directive it should work, but I don't think that's the best solution.

I updated the pen you made last time to reproduce the problem: https://codepen.io/guiiamorim/pen/YzvPXoY

**Reply by SimoTod** [reply]
You need to make sure the effect triggers when value changes as well. Just add all the variables you want to track in the value of the custom directive
`x-tail-select="[options, value]"`

**Reply by guiiamorim** [reply]
I thought so, was wondering if there was a better way to accomplish this, since I already have the `value` part on the `x-model="value"`

**Reply by SimoTod** [reply]
it depends on what you think it's better. Inside your directive, you have access to the element so you can check if the x-model attribute exists and get value from there: https://codepen.io/SimoTod/pen/gOKbLoX

**Reply by guiiamorim** [reply]
I think that's a better way, will try to use it like this. Thank you!
